File: ugoite-core/ugoite_core/audit.py
"""Audit logging primitives for security-relevant space events."""

from __future__ import annotations

import asyncio
import hashlib
import json
import os
import secrets
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

_DEFAULT_AUDIT_LIMIT = 100
_MAX_AUDIT_LIMIT = 500
_DEFAULT_AUDIT_RETENTION = 5000
_space_locks: dict[str, asyncio.Lock] = {}
_space_locks_guard = asyncio.Lock()


@dataclass(frozen=True)
class AuditEventInput:
    """Input payload for audit event persistence."""

    action: str
    actor_user_id: str
    outcome: str
    target_type: str | None = None
    target_id: str | None = None
    request_method: str | None = None
    request_path: str | None = None
    request_id: str | None = None
    metadata: dict[str, Any] | None = None


@dataclass(frozen=True)
class AuditListFilter:
    """Filter and pagination options for audit retrieval."""

    offset: int = 0
    limit: int = _DEFAULT_AUDIT_LIMIT
    action: str | None = None
    actor_user_id: str | None = None
    outcome: str | None = None


def _now_iso() -> str:
    return datetime.now(tz=UTC).isoformat().replace("+00:00", "Z")


def _retention_limit() -> int:
    raw = os.environ.get("UGOITE_AUDIT_RETENTION_MAX_EVENTS")
    if not isinstance(raw, str) or not raw.strip():
        return _DEFAULT_AUDIT_RETENTION
    try:
        parsed = int(raw)
    except ValueError:
        return _DEFAULT_AUDIT_RETENTION
    return max(100, parsed)


def _event_hash(payload: dict[str, Any], prev_hash: str) -> str:
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":"))
    material = f"{prev_hash}:{canonical}".encode()
    return hashlib.sha256(material).hexdigest()


def _verify_chain(events: list[dict[str, Any]]) -> None:
    prev_hash = "root"
    for event in events:
        candidate = dict(event)
        expected_hash = candidate.pop("event_hash", None)
        candidate_prev_hash = candidate.get("prev_hash", "root")
        if not isinstance(expected_hash, str):
            msg = "Audit event missing event_hash"
            raise TypeError(msg)
        if candidate_prev_hash != prev_hash:
            msg = "Audit chain prev_hash mismatch"
            raise RuntimeError(msg)
        actual_hash = _event_hash(candidate, prev_hash)
        if actual_hash != expected_hash:
            msg = "Audit chain integrity check failed"
            raise RuntimeError(msg)
        prev_hash = expected_hash


def _normalize_outcome(outcome: str) -> str:
    value = outcome.strip().lower()
    return value if value in {"success", "deny", "error"} else "success"


async def _space_lock(space_id: str) -> asyncio.Lock:
    async with _space_locks_guard:
        existing = _space_locks.get(space_id)
        if existing is not None:
            return existing
        created = asyncio.Lock()
        _space_locks[space_id] = created
        return created


def _audit_file_path(storage_config: dict[str, str], space_id: str) -> Path:
    uri = storage_config.get("uri", "")
    if not uri.startswith("fs://"):
        msg = "audit logging currently supports fs:// storage only"
        raise RuntimeError(msg)
    root = uri.removeprefix("fs://")
    base = Path(root)
    return base / "spaces" / space_id / "audit" / "events.jsonl"


async def _read_events(
    storage_config: dict[str, str],
    space_id: str,
) -> list[dict[str, Any]]:
    path = _audit_file_path(storage_config, space_id)
    if not path.exists():
        return []
    events: list[dict[str, Any]] = []
    text = await asyncio.to_thread(path.read_text, "utf-8")
    for line in text.splitlines():
        stripped = line.strip()
        if not stripped:
            continue
        try:
            parsed = json.loads(stripped)
        except json.JSONDecodeError as exc:
            msg = "Audit log contains malformed JSON"
            raise RuntimeError(msg) from exc
        if isinstance(parsed, dict):
            events.append(parsed)
    return events


async def _write_events(
    storage_config: dict[str, str],
    space_id: str,
    events: list[dict[str, Any]],
) -> None:
    path = _audit_file_path(storage_config, space_id)
    await asyncio.to_thread(path.parent.mkdir, parents=True, exist_ok=True)
    lines = [json.dumps(item, separators=(",", ":"), sort_keys=True) for item in events]
    payload = "\n".join(lines)
    if payload:
        payload += "\n"
    tmp_path = path.with_suffix(".tmp")
    await asyncio.to_thread(tmp_path.write_text, payload, "utf-8")
    await asyncio.to_thread(tmp_path.replace, path)


async def append_audit_event(
    storage_config: dict[str, str],
    space_id: str,
    payload: AuditEventInput,
) -> dict[str, Any]:
    """Append a tamper-evident audit event to space settings metadata."""
    action = payload.action.strip()
    if not action:
        msg = "audit action must not be empty"
        raise RuntimeError(msg)

    actor_user_id = payload.actor_user_id.strip()
    if not actor_user_id:
        msg = "actor_user_id must not be empty"
        raise RuntimeError(msg)

    lock = await _space_lock(space_id)
    async with lock:
        events = await _read_events(storage_config, space_id)
        _verify_chain(events)

        prev_hash = "root"
        if events:
            prev_hash_obj = events[-1].get("event_hash")
            if isinstance(prev_hash_obj, str) and prev_hash_obj:
                prev_hash = prev_hash_obj

        event: dict[str, Any] = {
            "id": secrets.token_urlsafe(12),
            "timestamp": _now_iso(),
            "space_id": space_id,
            "action": action,
            "actor_user_id": actor_user_id,
            "outcome": _normalize_outcome(payload.outcome),
            "target_type": payload.target_type,
            "target_id": payload.target_id,
            "request_method": payload.request_method,
            "request_path": payload.request_path,
            "request_id": payload.request_id,
            "metadata": payload.metadata or {},
            "prev_hash": prev_hash,
        }

        event["event_hash"] = _event_hash(event, prev_hash)
        events.append(event)

        retention = _retention_limit()
        if len(events) > retention:
            events = events[-retention:]

        await _write_events(storage_config, space_id, events)
        return event


async def list_audit_events(
    storage_config: dict[str, str],
    space_id: str,
    filters: AuditListFilter | None = None,
) -> dict[str, Any]:
    """List audit events with optional filters and pagination."""
    lock = await _space_lock(space_id)
    async with lock:
        all_events = await _read_events(storage_config, space_id)
        _verify_chain(all_events)

    options = filters or AuditListFilter()
    normalized_limit = max(1, min(options.limit, _MAX_AUDIT_LIMIT))
    normalized_offset = max(0, options.offset)
    normalized_action = (
        options.action.strip()
        if isinstance(options.action, str) and options.action.strip()
        else None
    )
    normalized_actor = (
        options.actor_user_id.strip()
        if isinstance(options.actor_user_id, str) and options.actor_user_id.strip()
        else None
    )
    normalized_outcome = (
        options.outcome.strip().lower()
        if isinstance(options.outcome, str) and options.outcome.strip()
        else None
    )

    filtered = all_events
    if normalized_action:
        filtered = [
            item for item in filtered if item.get("action") == normalized_action
        ]
    if normalized_actor:
        filtered = [
            item for item in filtered if item.get("actor_user_id") == normalized_actor
        ]
    if normalized_outcome:
        filtered = [
            item for item in filtered if item.get("outcome") == normalized_outcome
        ]

    filtered.sort(key=lambda item: str(item.get("timestamp", "")), reverse=True)
    total = len(filtered)
    page = filtered[normalized_offset : normalized_offset + normalized_limit]
    return {
        "items": page,
        "total": total,
        "offset": normalized_offset,
        "limit": normalized_limit,
    }

File: crates/ugoite-core/src/audit.rs
use anyhow::{anyhow, Result};
use chrono::{SecondsFormat, Utc};
use opendal::Operator;
use regex::Regex;
use serde_json::{json, Value};
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::sync::{Arc, OnceLock};
use tokio::sync::Mutex;

const DEFAULT_AUDIT_LIMIT: usize = 100;
const MAX_AUDIT_LIMIT: usize = 500;
const DEFAULT_AUDIT_RETENTION: usize = 5000;
const MAX_AUDIT_RETENTION: usize = 50000;

#[derive(Debug, Clone)]
pub struct AuditListOptions {
    pub offset: usize,
    pub limit: usize,
    pub action: Option<String>,
    pub actor_user_id: Option<String>,
    pub outcome: Option<String>,
}

impl Default for AuditListOptions {
    fn default() -> Self {
        Self {
            offset: 0,
            limit: DEFAULT_AUDIT_LIMIT,
            action: None,
            actor_user_id: None,
            outcome: None,
        }
    }
}

static SPACE_LOCKS: OnceLock<Mutex<HashMap<String, Arc<Mutex<()>>>>> = OnceLock::new();
static SPACE_ID_PATTERN: OnceLock<Regex> = OnceLock::new();

fn lock_registry() -> &'static Mutex<HashMap<String, Arc<Mutex<()>>>> {
    SPACE_LOCKS.get_or_init(|| Mutex::new(HashMap::new()))
}

async fn space_lock(space_id: &str) -> Arc<Mutex<()>> {
    let mut registry = lock_registry().lock().await;
    if let Some(existing) = registry.get(space_id) {
        return existing.clone();
    }
    let created = Arc::new(Mutex::new(()));
    registry.insert(space_id.to_string(), created.clone());
    created
}

fn normalize_retention_limit(limit: Option<usize>) -> usize {
    let raw = limit.unwrap_or(DEFAULT_AUDIT_RETENTION);
    raw.clamp(100, MAX_AUDIT_RETENTION)
}

fn normalize_outcome(outcome: Option<&str>) -> String {
    let normalized = outcome.unwrap_or("success").trim().to_lowercase();
    match normalized.as_str() {
        "success" | "deny" | "error" => normalized,
        _ => "success".to_string(),
    }
}

fn validate_space_id(space_id: &str) -> Result<String> {
    let normalized = space_id.trim();
    if normalized.is_empty() {
        return Err(anyhow!("space_id must not be empty"));
    }
    let pattern = SPACE_ID_PATTERN.get_or_init(|| {
        Regex::new(r"^[a-zA-Z0-9][a-zA-Z0-9._-]{0,127}$").expect("space id regex must be valid")
    });
    if !pattern.is_match(normalized) {
        return Err(anyhow!("invalid space_id"));
    }
    Ok(normalized.to_string())
}

fn audit_file_path(space_id: &str) -> String {
    format!("spaces/{space_id}/audit/events.jsonl")
}

fn now_iso() -> String {
    Utc::now().to_rfc3339_opts(SecondsFormat::Millis, true)
}

fn event_hash(payload: &Value, prev_hash: &str) -> Result<String> {
    let canonical = serde_json::to_string(payload)?;
    let material = format!("{prev_hash}:{canonical}");
    let digest = Sha256::digest(material.as_bytes());
    Ok(hex::encode(digest))
}

fn verify_chain(events: &[Value]) -> Result<()> {
    let mut prev_hash = "root".to_string();
    for event in events {
        let mut candidate = event.clone();
        let object = candidate
            .as_object_mut()
            .ok_or_else(|| anyhow!("Audit log contains malformed JSON"))?;
        let expected_hash = object
            .remove("event_hash")
            .and_then(|v| v.as_str().map(str::to_string))
            .ok_or_else(|| anyhow!("Audit event missing event_hash"))?;
        let candidate_prev_hash = object
            .get("prev_hash")
            .and_then(Value::as_str)
            .unwrap_or("root");
        if candidate_prev_hash != prev_hash {
            return Err(anyhow!("Audit chain prev_hash mismatch"));
        }
        let actual_hash = event_hash(&candidate, &prev_hash)?;
        if actual_hash != expected_hash {
            return Err(anyhow!("Audit chain integrity check failed"));
        }
        prev_hash = expected_hash;
    }
    Ok(())
}

fn rehash_chain(events: &mut [Value]) -> Result<()> {
    let mut prev_hash = "root".to_string();
    for event in events.iter_mut() {
        {
            let object = event
                .as_object_mut()
                .ok_or_else(|| anyhow!("Audit log contains malformed JSON"))?;
            object.insert("prev_hash".to_string(), Value::String(prev_hash.clone()));
            object.remove("event_hash");
        }
        let hash = event_hash(event, &prev_hash)?;
        let object = event
            .as_object_mut()
            .ok_or_else(|| anyhow!("Audit log contains malformed JSON"))?;
        object.insert("event_hash".to_string(), Value::String(hash.clone()));
        prev_hash = hash;
    }
    Ok(())
}

async fn read_events(op: &Operator, space_id: &str) -> Result<Vec<Value>> {
    let path = audit_file_path(space_id);
    if !op.exists(&path).await? {
        return Ok(Vec::new());
    }
    let bytes = op.read(&path).await?;
    let content = String::from_utf8(bytes.to_vec())?;
    let mut events = Vec::new();
    for line in content.lines() {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }
        let parsed: Value = serde_json::from_str(trimmed)
            .map_err(|_| anyhow!("Audit log contains malformed JSON"))?;
        if parsed.is_object() {
            events.push(parsed);
        }
    }
    Ok(events)
}

async fn write_events(op: &Operator, space_id: &str, events: &[Value]) -> Result<()> {
    let dir_path = format!("spaces/{space_id}/audit/");
    op.create_dir(&dir_path).await?;
    let path = audit_file_path(space_id);
    let mut lines = Vec::with_capacity(events.len());
    for item in events {
        lines.push(serde_json::to_string(item)?);
    }
    let mut payload = lines.join("\n");
    if !payload.is_empty() {
        payload.push('\n');
    }
    op.write(&path, payload.into_bytes()).await?;
    Ok(())
}

pub async fn append_audit_event(
    op: &Operator,
    space_id: &str,
    payload: &Value,
    retention_limit: Option<usize>,
) -> Result<Value> {
    let safe_space_id = validate_space_id(space_id)?;
    let payload_obj = payload
        .as_object()
        .ok_or_else(|| anyhow!("audit payload must be an object"))?;

    let action = payload_obj
        .get("action")
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .ok_or_else(|| anyhow!("audit action must not be empty"))?
        .to_string();

    let actor_user_id = payload_obj
        .get("actor_user_id")
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .ok_or_else(|| anyhow!("actor_user_id must not be empty"))?
        .to_string();

    let lock = space_lock(&safe_space_id).await;
    let _guard = lock.lock().await;

    let mut events = read_events(op, &safe_space_id).await?;
    verify_chain(&events)?;

    let prev_hash = events
        .last()
        .and_then(Value::as_object)
        .and_then(|item| item.get("event_hash"))
        .and_then(Value::as_str)
        .unwrap_or("root")
        .to_string();

    let metadata = payload_obj
        .get("metadata")
        .and_then(Value::as_object)
        .map(|_| {
            payload_obj
                .get("metadata")
                .cloned()
                .unwrap_or_else(|| json!({}))
        })
        .unwrap_or_else(|| json!({}));

    let mut event = json!({
        "id": format!("audit-{}", uuid::Uuid::new_v4().simple()),
        "timestamp": now_iso(),
        "space_id": safe_space_id,
        "action": action,
        "actor_user_id": actor_user_id,
        "outcome": normalize_outcome(payload_obj.get("outcome").and_then(Value::as_str)),
        "target_type": payload_obj.get("target_type").cloned().unwrap_or(Value::Null),
        "target_id": payload_obj.get("target_id").cloned().unwrap_or(Value::Null),
        "request_method": payload_obj.get("request_method").cloned().unwrap_or(Value::Null),
        "request_path": payload_obj.get("request_path").cloned().unwrap_or(Value::Null),
        "request_id": payload_obj.get("request_id").cloned().unwrap_or(Value::Null),
        "metadata": metadata,
        "prev_hash": prev_hash,
    });

    let hash = event_hash(&event, event["prev_hash"].as_str().unwrap_or("root"))?;
    event["event_hash"] = Value::String(hash);
    events.push(event.clone());

    let retention = normalize_retention_limit(retention_limit);
    if events.len() > retention {
        let start_index = events.len() - retention;
        events = events.split_off(start_index);
        rehash_chain(&mut events)?;
        if let Some(last) = events.last() {
            event = last.clone();
        }
    }

    write_events(op, &safe_space_id, &events).await?;
    Ok(event)
}

pub async fn list_audit_events(
    op: &Operator,
    space_id: &str,
    options: AuditListOptions,
) -> Result<Value> {
    let safe_space_id = validate_space_id(space_id)?;
    let lock = space_lock(&safe_space_id).await;
    let _guard = lock.lock().await;

    let mut events = read_events(op, &safe_space_id).await?;
    verify_chain(&events)?;

    let action = options
        .action
        .as_deref()
        .map(str::trim)
        .filter(|value| !value.is_empty());
    let actor = options
        .actor_user_id
        .as_deref()
        .map(str::trim)
        .filter(|value| !value.is_empty());
    let outcome = options
        .outcome
        .as_deref()
        .map(str::trim)
        .map(str::to_lowercase)
        .filter(|value| !value.is_empty());

    events.retain(|event| {
        let Some(obj) = event.as_object() else {
            return false;
        };
        if let Some(action_value) = action {
            if obj.get("action").and_then(Value::as_str) != Some(action_value) {
                return false;
            }
        }
        if let Some(actor_value) = actor {
            if obj.get("actor_user_id").and_then(Value::as_str) != Some(actor_value) {
                return false;
            }
        }
        if let Some(ref outcome_value) = outcome {
            if obj.get("outcome").and_then(Value::as_str) != Some(outcome_value.as_str()) {
                return false;
            }
        }
        true
    });

    events.sort_by(|left, right| {
        let left_ts = left
            .as_object()
            .and_then(|obj| obj.get("timestamp"))
            .and_then(Value::as_str)
            .unwrap_or("");
        let right_ts = right
            .as_object()
            .and_then(|obj| obj.get("timestamp"))
            .and_then(Value::as_str)
            .unwrap_or("");
        right_ts.cmp(left_ts)
    });

    let normalized_limit = options.limit.clamp(1, MAX_AUDIT_LIMIT);
    let normalized_offset = options.offset;
    let total = events.len();
    let items: Vec<Value> = events
        .into_iter()
        .skip(normalized_offset)
        .take(normalized_limit)
        .collect();

    Ok(json!({
        "items": items,
        "total": total,
        "offset": normalized_offset,
        "limit": normalized_limit,
    }))
}

pub fn default_retention_from_env() -> usize {
    let parsed = std::env::var("UGOITE_AUDIT_RETENTION_MAX_EVENTS")
        .ok()
        .and_then(|raw| raw.trim().parse::<usize>().ok());
    normalize_retention_limit(parsed)
}
